sympy/solvers/ode.py

```diff
@@ -5755,6 +5755,11 @@ def _solve_undetermined_coefficients(eq, func, order, match):
 
     newtrialset = trialset - notneedset
 
+    newtrialset = list(newtrialset)
+    while simplify(wronskian(newtrialset+gensols,x))== 0:
+        for i in range(len(newtrialset)):
+            newtrialset[i] *= x
```

Comment:
As the trialset grows the calculating the wronskian scales very badly so this can be very slow. Try for example:
```
dsolve(f(x).diff(x, 2) + f(x) - x**4*sin(x-1))
```
In that case the trial-set has size 10 so for the wronskian we have to compute the determinant of a 10x10 symbolic matrix. Calculating a determinant is roughly `O(n**3)` for nxn matrices and gives as many terms so this will generate a 1000 term wronskian that is then to be simplified.

Also we normally try to avoid using general simplify in library code because it can be slow. It is usually better to use a particular kind of simplification that you expect to be useful. We have a pretty good idea what sort of functions we are trying to simplify here so we should be able to come up with something more specific.

I suggest checking to see if any of the solutions in the trialset is a solution of the homogeneous part of the equation though. That is linear in the size of the trialset and doesn't need to involve simplify etc.

The code here would look like:
```
    # If any element of the trial set solves the homogeneous part of the ODE
    # then every element of the trial set needs to be multiplied by x
    while any(sub_func_doit(eq_hom, f(x), ts).is_zero for ts in trialset):
        trialset = {x*ts for ts in trialset}
```
The only problem is that this function doesn't have access to the ODE or to its homogeneous part. That must have been calculated elsewhere though so we just need to go back to where `match` is generated and store these things there.

Reply:
so the code which i have written should be replaced by your code block?

> The only problem is that this function doesn't have access to the ODE or to its homogeneous part. That must have been calculated elsewhere though so we just need to go back to where match is generated and store these things there.

sorry,but i didn't get this

Reply:
Actually you do have access to the eq here. You need to split the equation into the homogeneous and non-homogeneous part e.g.:
```
In [1]: eq = f(x).diff(x, 2) + f(x) + sin(x)                                                                                                   

In [2]: eq                                                                                                                                     
Out[2]: 
                  2      
                 d       
f(x) + sin(x) + ───(f(x))
                  2      
                dx       

In [3]: eq_hom = f(x).diff(x, 2) + f(x)                                                                                                        

In [4]: eq_hom                                                                                                                                 
Out[4]: 
         2      
        d       
f(x) + ───(f(x))
         2      
       dx  
```

Reply:
i tried to separate but there may be lot of corner cases which may fail so could you please share approach for separation of non homogenuos part

Reply:
There must be code that already does this somewhere.

Reply:
i tried to look for it but couldn't find. `Chareq` is made for finding complimentory solutions

Reply:
```julia
In [1]: from sympy.solvers.ode import _nth_linear_match                                                                                        

In [2]: eq = f(x).diff(x, 2) + f(x) - sin(x)                                                                                                   

In [3]: eq                                                                                                                                     
Out[3]: 
                  2      
                 d       
f(x) - sin(x) + ───(f(x))
                  2      
                dx       

In [4]: _nth_linear_match(eq, f(x), 2)                                                                                                         
Out[4]: {-1: -sin(x), 0: 1, 1: 0, 2: 1}
```

Reply:
There is another problem with this approach:
```julia
In [3]: dsolve(f(x).diff(x, 2)+f(x)-(sin(x-2)+1))                                                                                              
---------------------------------------------------------------------------
KeyError
~/current/sympy/sympy/sympy/solvers/ode.py in _solve_undetermined_coefficients(eq, func, order, match)
   5781     for i in Add.make_args(eqs):
   5782         s = separatevars(i, dict=True, symbols=[x])
-> 5783         coeffsdict[s[x]] += s['coeff']
   5784 
   5785     coeffvals = solve(list(coeffsdict.values()), coefflist)

KeyError: x
```
The reason this doesn't work is that the trialset is
```
[1, cos(x - 2), sin(x - 2)]
```
After checking the Wronskian these are all multiplied by `x` giving
```
[x, x*cos(x - 2), x*sin(x - 2)]
```
However the `1` should not have been multiplied by `x`. We should multiply each member of a family *before* we take the union of the families for each added term.

Reply:
so we should check this while loop inside trialset function.

Reply:
if we do that then a lot of changes may need to done like passing extra arguments

Reply:
That would be the simplest thing. I think ideally we should move most of the code for building the trial set from the matching phase to the solving phase if possible.

Reply:
@oscarbenjamin i have tried to change the trialset generation function but now it still shows key error for `sin(x-1)` in the example `dsolve(f(x).diff(x, 2)+f(x)-sin(x-1))`  whereas the trialset returned by the function is now perfect.

Reply:
```
assert _undetermined_coefficients_match(
        x**2*sin(x)*exp(x) + x*sin(x) + x, x
    ) == {
        'test': True, 'trialset': set([x**2*cos(x)*exp(x), x, cos(x), S.One,
        exp(x)*sin(x), sin(x), x*exp(x)*sin(x), x*cos(x), x*cos(x)*exp(x),
        x*sin(x), cos(x)*exp(x), x**2*exp(x)*sin(x)])}
```
this test case is failing but when i tested the coded it shows correct result. so can you please help me in this?